### PR TITLE
Vault Zaps

### PR DESCRIPTION
**Pull request by scooprinder**
# Description
* Adds initial skeleton of the AscendZaps contract
* Removed any faith related code from the VaultProxy 

# Checklist
- [ ] Code follows the style guide
- [ ] I have performed a self-review of my own code
- [ ] New and existing tests pass locally
- [ ] This PR is targeting the correct branch

### CODE REVIEW

File: protocol/test/core/ascend-zaps-test.ts [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+  let joiningFee: JoiningFee;
+  let vaultedTemple: VaultedTemple;
+  let ascendZaps: AscendZaps;
+  let templeExposure: Exposure;
+
+  let owner: Signer;
+  let alan: SignerWithAddress;
+  let ben: Signer;
+  const periodDuration = 60 * 10;
+  const balanceVaultAddress = '0xBA12222222228d8Ba445958a75a0704d566BF2C8';
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:44]: keep constants in `constants.ts`

File: protocol/test/core/ascend-zaps-test.ts [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+            blockNumber: 15674900,
+          },
+        },
+      ],
+    });
+
+    [owner, alan, ben] = await ethers.getSigners();
+
+    TEMPLE = await TempleERC20Token__factory.connect(
+      '0x470EBf5f030Ed85Fc1ed4C2d36B9DD02e77CF1b7',
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:65]: import from `constants.ts`

File: protocol/test/core/ascend-zaps-test.ts [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+import { expect } from 'chai';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
+import { advanceTimeAndBlock } from '../utils/time';
+import { impersonateAccount } from '../utils/account';
+import { takeSnapshot, revertToSnapshot } from '../utils/snapshot';
+import constants from '../constants';
+
+config();
+
+describe.only('Ascend Zaps', async () => {
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:31]: please remove `only`

File: protocol/test/utils/time.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import { ethers } from 'hardhat';
```
Comment (princetonbishop) [inline protocol/test/utils/time.ts:1]: `helpers.ts` already has a lot of these helper functions. please use those

File: protocol/test/utils/account.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { network } from 'hardhat';
```
Comment (princetonbishop) [inline protocol/test/utils/account.ts:1]: given it's only one function, I would suggest adding to `helpers.ts`

File: protocol/test/core/ascend-zaps-test.ts [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+        ascendZaps.redeemVaultTokenToTemple(vault.address)
+      ).to.be.revertedWith('CanNotExitVault');
+
+      await advanceTimeAndBlock(periodDuration);
+
+      await ascendZaps.redeemVaultTokenToTemple(vault.address);
+    });
+
+    it('only owner can redeem', async () => {
+      await advanceTimeAndBlock(periodDuration);
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:413]: use `mineToTimestamp` from `helpers.ts`

File: protocol/test/core/ascend-zaps-test.ts [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+
+      await vault.connect(alan).approve(ascendZaps.address, swapAmount);
+      await ascendZaps
+        .connect(alan)
+        .exitVaultEarly(
+          swapAmount,
+          vault.address,
+          singleSwap,
+          fundMng,
+          0,
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:439]: I suggest you get swap limit from balancer and check user has at least that additional amount after `exitVaultEarly`

File: protocol/test/core/ascend-zaps-test.ts [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+      const beforeDAIBalance = await DAI.balanceOf(alan.address);
+
+      const amountCalculated = await ascendZaps
+        .connect(alan)
+        .callStatic.exitVaultEarly(
+          swapAmount,
+          vault.address,
+          singleSwap,
+          fundMng,
+          0,
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:340]: get swap limits from balancer and verify user has at least that many additional tokens.

File: protocol/test/core/ascend-zaps-test.ts [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+          deadline
+        );
+
+      interestAccrued = await ascendZaps.checkInterestAccrued(vault.address);
+      expect(interestAccrued).to.be.eq(0);
+
+      await ascendZaps.redeemVaultTokenToTemple(vault.address);
+      interestAccrued = await ascendZaps.checkInterestAccrued(vault.address);
+      expect(interestAccrued).to.be.eq(0);
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:448]: simulate extra transactions to make sure interest is accrued. should get some hints in other test files.

File: protocol/test/core/ascend-zaps-test.ts [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+      it('withdraw zero amount', async () => {
+        expect(
+          ascendZaps
+            .connect(owner)
+            .withdraw(
+              ethers.constants.AddressZero,
+              await alan.getAddress(),
+              toAtto(0)
+            )
+        ).to.be.revertedWith('WithdrawZeroAmount');
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:528]: try `revertedWithCustomError`

File: protocol/test/core/ascend-zaps-test.ts [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+
+      it('withdraw up to balance', async () => {
+        const beforeBalance = await waffle.provider.getBalance(alan.address);
+        await ascendZaps
+          .connect(owner)
+          .withdraw(ethers.constants.AddressZero, alan.address, amount);
+
+        const afterBalance = await waffle.provider.getBalance(alan.address);
+        expect(await waffle.provider.getBalance(ascendZaps.address)).to.be.eq(
+          0
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:551]: nit: move to same line

File: protocol/contracts/zaps/AscendZaps.sol [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    uint256 currentShareAmount = vaultErc.toSharesAmount(balance);
+    uint256 currentTempleAmount = vaultErc.toTokenAmount(currentShareAmount);
+
+    uint256 delta = withdrawn[address(vaultErc)] - currentTempleAmount;
+    delete withdrawn[address(vaultErc)];
+
+    vaultErc.withdraw(balance);
+
+    // redistribute interest
+    temple.safeTransfer(vaultedTemple, delta);
```
Comment (princetonbishop) [inline protocol/contracts/zaps/AscendZaps.sol:138]: check enough temple balance to transfer

File: protocol/test/core/ascend-zaps-test.ts [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+      periodDuration,
+      60,
+      { p: 1, q: 1 },
+      joiningFee.address,
+      await blockTimestamp()
+    );
+
+    await templeExposure.setMinterState(vault.address, true);
+
+    await expect(
```
Comment (princetonbishop) [inline protocol/test/core/ascend-zaps-test.ts:104]: you can remove these zero checks for constructor arguments. I also suggest you remove zero address checks from solidity contract constructor.

**Review by princetonbishop — CHANGES_REQUESTED** [review]
Good job, although I think there's room for some improvements. added some comments. Also:
- you would learn some stuff from looking at some other tests in the repo
- run slither (https://github.com/TempleDAO/temple/pull/579)

File: protocol/contracts/zaps/AscendZaps.sol [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+  event EarlyWithdraw(
+    address account,
+    uint256 fromAmount,
+    address vault,
+    address toToken,
+    uint256 toTokenAmount
+  );
+
+  constructor(address _vaultedTemple, IERC20 _temple) {
+    if (_vaultedTemple == address(0)) {
```
[the remark this answers is not in the document]
Comment (0xlock) [reply protocol/contracts/zaps/AscendZaps.sol:48]: that's right
but if we are using scripts to deploy contracts, we can't 100% sure correct value will always be passed to constructor
I think keeping it as is now is not harmful to the contract
not sure why you want to remove it

File: protocol/contracts/zaps/AscendZaps.sol [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    // Transfer vault tokens to contract, and determine how much Temple that is currently worth
+    IERC20(vaultErc).safeTransferFrom(msg.sender, address(this), amount);
+    uint256 shareAmount = vaultErc.toSharesAmount(amount);
+    uint256 templeBal = vaultErc.toTokenAmount(shareAmount);
+
+    // Check that the supplied offchain swap data is correct; not strictly needed
+    if (
+      address(singleSwap.assetIn) != address(temple) ||
+      singleSwap.amount != templeBal
```
[the remark this answers is not in the document]
Comment (0xlock) [reply protocol/contracts/zaps/AscendZaps.sol:89]: ok

File: protocol/contracts/zaps/AscendZaps.sol [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+    uint256 currentTempleAmount = vaultErc.toTokenAmount(currentShareAmount);
+
+    uint256 delta = withdrawn[address(vaultErc)] - currentTempleAmount;
+    delete withdrawn[address(vaultErc)];
+
+    vaultErc.withdraw(balance);
+
+    // redistribute interest
+    temple.safeTransfer(vaultedTemple, delta);
+  }
```
[the remark this answers is not in the document]
Comment (0xlock) [reply protocol/contracts/zaps/AscendZaps.sol:146]: ok

File: protocol/test/utils/time.ts (hunk 4 above)
Comment (0xlock) [reply protocol/test/utils/time.ts]: ok

File: protocol/test/utils/account.ts (hunk 5 above)
Comment (0xlock) [reply protocol/test/utils/account.ts]: ok

File: protocol/test/core/ascend-zaps-test.ts (hunk 6 above)
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts]: done!

File: protocol/test/core/ascend-zaps-test.ts (hunk 11 above)
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts:585]: ok

File: protocol/test/core/ascend-zaps-test.ts [hunk 17]
```diff
@@ -0,0 +1,557 @@
...
+          .connect(alan)
+          .exitVaultEarly(
+            swapAmount,
+            vault.address,
+            singleSwap,
+            fundMng,
+            0,
+            deadline
+          )
+      ).to.be.revertedWith('FirstVaultCycle');
```
[the remark this answers is not in the document]
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts]: done!

File: protocol/test/core/ascend-zaps-test.ts (hunk 10 above)
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts]: done!

File: protocol/test/core/ascend-zaps-test.ts [hunk 18]
```diff
@@ -0,0 +1,557 @@
...
+  let owner: Signer;
+  let alan: SignerWithAddress;
+  let ben: Signer;
+  const periodDuration = 60 * 10;
+  const balanceVaultAddress = '0xBA12222222228d8Ba445958a75a0704d566BF2C8';
+
+  let snapshotId: string;
+
+  before(async function () {
+    this.timeout(0);
```
[the remark this answers is not in the document]
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts]: correct, we don't need it

File: protocol/contracts/zaps/AscendZaps.sol (hunk 12 above)
Comment (0xlock) [reply protocol/contracts/zaps/AscendZaps.sol:140]: ok

File: protocol/test/core/ascend-zaps-test.ts (hunk 7 above)
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts:536]: done!

File: protocol/test/core/ascend-zaps-test.ts (hunk 8 above)
Comment (0xlock) [reply protocol/test/core/ascend-zaps-test.ts:438]: done!

**Pull request closed without merging**